Build failure: ceph

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

- Apply #425900.
- `nix-build -A ceph`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.ceph.x86_64-linux

### Relevant log output

```console
error: builder for '/nix/store/rpfwksnixm3kvi2wwg7zyaan7j763a1m-arrow-cpp-19.0.1.drv' failed with exit code 1;
       last 25 log lines:
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_strings_internal.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_strings_internal.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_string_view.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_string_view.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_base.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_base.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_spinlock_wait.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_spinlock_wait.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_throw_delegate.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_throw_delegate.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_raw_logging_internal.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_raw_logging_internal.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: warning: libabsl_log_severity.so.2505.0.0, needed by /nix/store/cx5fzs0swnfvavbdvi1zg1f4yfll0wnq-grpc-1.73.1/lib/libgrpc++.so.1.73.1, may conflict with libabsl_log_severity.so.2407.0.0
       > /nix/store/x7w0rmkaam7aikclyhwdbn8wmmq29ma6-binutils-2.44/bin/ld: release/libarrow_flight.so.1900.1.0: undefined reference to `grpc_call_run_in_event_engine(grpc_call const*, absl::lts_20240722::AnyInvocable<void ()>)'
       > collect2: error: ld returned 1 exit status
       > [451/845] Building CXX object src/arrow/testing/CMakeFiles/arrow_filesystem_example.dir/examplefs.cc.o
       > [452/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/extension_set.cc.o
       > [453/845] Building CXX object src/arrow/flight/sql/CMakeFiles/arrow_flight_sql_objlib.dir/client.cc.o
       > [454/845] Building CXX object src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/scanner.cc.o
       > [455/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/type_internal.cc.o
       > [456/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/plan_internal.cc.o
       > [457/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/test_plan_builder.cc.o
       > [458/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/util_internal.cc.o
       > [459/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/extended_expression_internal.cc.o
       > [460/845] Building CXX object src/arrow/flight/sql/CMakeFiles/arrow_flight_sql_objlib.dir/server.cc.o
       > [461/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/serde.cc.o
       > [462/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/options.cc.o
       > [463/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/relation_internal.cc.o
       > [464/845] Building CXX object src/arrow/engine/CMakeFiles/arrow_substrait_objlib.dir/substrait/expression_internal.cc.o
       > [465/845] Building CXX object src/arrow/CMakeFiles/arrow-array-test.dir/array/array_test.cc.o
       > ninja: build stopped: subcommand failed.
       For full logs, run:
         nix log /nix/store/rpfwksnixm3kvi2wwg7zyaan7j763a1m-arrow-cpp-19.0.1.drv
error: 1 dependencies of derivation '/nix/store/balz7133x5hsm7hwwwv4v1nblj5gldn1-ceph-19.2.2.drv' failed to build
```

### Additional context

#425732 must be resolved before this.

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.15.6-cachyos, NixOS, 25.11 (Xantusia), 25.11.20250708.9807714`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - nixpkgs: `/nix/store/bgl6ldj5ihbwcq8p42z3a0qzgqafgk2b-source`


### Notify maintainers

@adevress @alexanderkjeldaas @johanot @krav @nh2 @benaryorg
@aneeshusa This caused the build failure of `samba4Full`.

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

I'm still having issues with building even after reverting https://github.com/NixOS/nixpkgs/commit/5240499fb672c5f7e25cd84bff780062cb3cf75a and cherry-picking https://github.com/NixOS/nixpkgs/pull/427905

Getting some weird compiler segfault
```
arrow-cpp> [585/845] Building CXX object src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/bit_util_test.cc.o
arrow-cpp> FAILED: src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/bit_util_test.cc.o 
arrow-cpp> /nix/store/7nlf5v84s4p2yhx327j8495yik60qnzh-gcc-wrapper-14.3.0/bin/g++ -DARROW_EXTRA_ERROR_CONTEXT -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_WITH_TIMING_TESTS -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -I/build/source/cpp/build/src -I/build/source/cpp/src -I/build/source/cpp/src/generated -isystem /build/source/cpp/thirdparty/flatbuffers/include -Wno-noexcept-type -Wno-self-move  -fdiagnostics-color=always  -Wall -fno-semantic-interposition -msse4.2  -O3 -DNDEBUG -O2 -ftree-vectorize  -std=c++17 -fPIE -MD -MT src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/bit_util_test.cc.o -MF src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/bit_util_test.cc.o.d -o src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/bit_util_test.cc.o -c /build/source/cpp/src/arrow/util/bit_util_test.cc
arrow-cpp> during GIMPLE pass: ehcleanup
arrow-cpp> /build/source/cpp/src/arrow/util/bit_util_test.cc: In member function 'virtual void arrow::FirstTimeBitmapWriter_AppendWordOffsetOverwritesCorrectBitsOnExistingByte_Test::TestBody()':
arrow-cpp> /build/source/cpp/src/arrow/util/bit_util_test.cc:2369:1: internal compiler error: Segmentation fault
arrow-cpp>  2369 | }  // namespace arrow
arrow-cpp>       | ^
arrow-cpp> 0x21b24aa diagnostic_impl(rich_location*, diagnostic_metadata const*, int, char const*, __va_list_tag (*) [1], diagnostic_t)
arrow-cpp>      ???:0
arrow-cpp> 0x21b2ca2 internal_error(char const*, ...)
arrow-cpp>      ???:0
arrow-cpp> 0xfb5247 crash_signal(int)
arrow-cpp>      ???:0
arrow-cpp> g++: internal compiler error: Segmentation fault signal terminated program cc1plus
arrow-cpp> Please submit a full bug report, with preprocessed source (by using -freport-bug).
arrow-cpp> See <https://gcc.gnu.org/bugs/> for instructions.
arrow-cpp> [586/845] Building CXX object src/arrow/util/CMakeFiles/arrow-threading-utility-test.dir/counting_semaphore_test.cc.o
arrow-cpp> [587/845] Building CXX object src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_util_test.cc.o
arrow-cpp> [588/845] Building CXX object src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_generator_test.cc.o
arrow-cpp> FAILED: src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_generator_test.cc.o 
arrow-cpp> /nix/store/7nlf5v84s4p2yhx327j8495yik60qnzh-gcc-wrapper-14.3.0/bin/g++ -DARROW_EXTRA_ERROR_CONTEXT -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_WITH_TIMING_TESTS -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -I/build/source/cpp/build/src -I/build/source/cpp/src -I/build/source/cpp/src/generated -isystem /build/source/cpp/thirdparty/flatbuffers/include -Wno-noexcept-type -Wno-self-move  -fdiagnostics-color=always  -Wall -fno-semantic-interposition -msse4.2  -O3 -DNDEBUG -O2 -ftree-vectorize  -std=c++17 -fPIE -MD -MT src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_generator_test.cc.o -MF src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_generator_test.cc.o.d -o src/arrow/util/CMakeFiles/arrow-async-utility-test.dir/async_generator_test.cc.o -c /build/source/cpp/src/arrow/util/async_generator_test.cc
arrow-cpp> during GIMPLE pass: fixup_cfg
arrow-cpp> /build/source/cpp/src/arrow/util/async_generator_test.cc: In member function 'virtual void arrow::MergedGeneratorTestFixture_MergedInnerFail_Test::TestBody()':
arrow-cpp> /build/source/cpp/src/arrow/util/async_generator_test.cc:1965:1: internal compiler error: Segmentation fault
arrow-cpp>  1965 | }  // namespace arrow
arrow-cpp>       | ^
arrow-cpp> 0x21b24aa diagnostic_impl(rich_location*, diagnostic_metadata const*, int, char const*, __va_list_tag (*) [1], diagnostic_t)
arrow-cpp>      ???:0
arrow-cpp> 0x21b2ca2 internal_error(char const*, ...)
arrow-cpp>      ???:0
arrow-cpp> 0xfb5247 crash_signal(int)
arrow-cpp>      ???:0
arrow-cpp> g++: internal compiler error: Segmentation fault signal terminated program cc1plus
arrow-cpp> Please submit a full bug report, with preprocessed source (by using -freport-bug).
arrow-cpp> See <https://gcc.gnu.org/bugs/> for instructions.
arrow-cpp> [589/845] Building CXX object src/arrow/util/CMakeFiles/arrow-utility-test.dir/decimal_test.cc.o
arrow-cpp> [590/845] Building CXX object src/arrow/util/CMakeFiles/arrow-bit-utility-test.dir/rle_encoding_test.cc.o
arrow-cpp> [591/845] Building CXX object src/arrow/util/CMakeFiles/arrow-threading-utility-test.dir/task_group_test.cc.o
arrow-cpp> [592/845] Building CXX object src/arrow/util/CMakeFiles/arrow-threading-utility-test.dir/future_test.cc.o
arrow-cpp> [593/845] Building CXX object src/arrow/compute/kernels/CMakeFiles/arrow-compute-scalar-if-else-test.dir/scalar_if_else_test.cc.o
arrow-cpp> [594/845] Building CXX object src/arrow/util/CMakeFiles/arrow-utility-test.dir/small_vector_test.cc.o
arrow-cpp> [595/845] Building CXX object src/arrow/compute/kernels/CMakeFiles/arrow-compute-vector-test.dir/vector_replace_test.cc.o
arrow-cpp> [596/845] Building CXX object src/arrow/compute/kernels/CMakeFiles/arrow-compute-aggregate-test.dir/aggregate_test.cc.o
arrow-cpp> ninja: build stopped: subcommand failed.
```

--- a/pkgs/by-name/sn/snappy/package.nix
+++ b/pkgs/by-name/sn/snappy/package.nix
@@ -3,6 +3,7 @@
   stdenv,
   fetchFromGitHub,
   cmake,
+  fetchpatch,
   static ? stdenv.hostPlatform.isStatic,
 }:
 
@@ -19,6 +20,22 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./revert-PUBLIC.patch
+    # Re-enable RTTI, without which other applications can't subclass snappy::Source
+    # While the patch was rejected upstream, it does not make it any less necessary to carry forward.
+    # ==> lack of RTTI *breaks* Ceph (and others) <==
+    #
+    # https://tracker.ceph.com/issues/53060
+    # https://build.opensuse.org/package/show/openSUSE:Factory/snappy
+    #
+    # Should this patch fail to apply use the above site to get the updated patch (rev in the url below).
+    # On the page there's a "latest revision" section which lists the last request which was merged into it.
+    # Click the "Request <insert number>" link, then view any file using "View file", and copy the rev from your address bar.
+    # For a different revision (in case nixpkgs is behind or something) you can go through the full revision history.
+    # Should the patch not be available for the nixpkgs version, ideally wait until the patch becomes available before bumping, or vendor it if necessary.
+    (fetchpatch {
+      url = "https://build.opensuse.org/public/source/openSUSE:Factory/snappy/reenable-rtti.patch?rev=e3449869b466869fc6b8a03a1a528fa6";
+      hash = "sha256-JhVhkHh7XPx1Bzf5xnOgWLgwh1oihX3O+emQWzE4Dho=";
+    })
   ];
 
   outputs = [

--- a/pkgs/tools/filesystems/ceph/arrow-cpp-19.nix
+++ b/pkgs/tools/filesystems/ceph/arrow-cpp-19.nix
@@ -8,6 +8,7 @@
   stdenv,
   lib,
   fetchurl,
+  fetchpatch2,
   fetchFromGitHub,
   fixDarwinDylibNames,
   autoconf,
@@ -42,7 +43,7 @@
   openssl,
   perl,
   pkg-config,
-  protobuf_29,
+  protobuf,
   python3,
   rapidjson,
   re2,
@@ -67,9 +68,6 @@
 }:
 
 let
-  # https://github.com/apache/arrow/issues/45807
-  protobuf = protobuf_29;
-
   arrow-testing = fetchFromGitHub {
     name = "arrow-testing";
     owner = "apache";
@@ -100,6 +98,15 @@ stdenv.mkDerivation (finalAttrs: {
   };
 
   sourceRoot = "${finalAttrs.src.name}/cpp";
+
+  patches = [
+    (fetchpatch2 {
+      name = "protobuf-30-compat.patch";
+      url = "https://github.com/apache/arrow/pull/46136.patch";
+      hash = "sha256-WTpe/eT3himlCHN/R78w1sF0HG859mE2ZN70U+9N8Ag=";
+      stripLen = 1;
+    })
+  ];
 
   # versions are all taken from
   # https://github.com/apache/arrow/blob/apache-arrow-${version}/cpp/thirdparty/versions.txt

